File: app/src/main/java/com/example/tugis3/gnss/NmeaParser.java
package com.example.tugis3.gnss;

import android.util.Log;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class NmeaParser {
    private static final String TAG = "NmeaParser";

    public static class Sentence {
        public String raw;
        public String talker;
        public String type;
        public List<String> fields = new ArrayList<>();
        public String checksum;

        @Override
        public String toString() {
            return "Sentence{" +
                    "talker='" + talker + '\'' +
                    ", type='" + type + '\'' +
                    ", fields=" + fields +
                    ", checksum='" + checksum + '\'' +
                    '}';
        }
    }

    public static Sentence parse(String line) {
        if (line == null) return null;
        line = line.trim();
        if (line.isEmpty()) return null;
        if (!line.startsWith("$")) return null;
        Sentence s = new Sentence();
        s.raw = line;
        try {
            String body, chk = null;
            int asterisk = line.indexOf('*');
            if (asterisk >= 0) {
                body = line.substring(1, asterisk);
                chk = line.substring(asterisk + 1);
            } else {
                body = line.substring(1);
            }
            s.checksum = chk;
            String[] parts = body.split(",");
            if (parts.length > 0) {
                String id = parts[0];
                if (id.length() >= 5) {
                    s.talker = id.substring(0, 2);
                    s.type = id.substring(2);
                } else if (id.length() >= 3) {
                    s.talker = id.substring(0, 2);
                    s.type = id.substring(2);
                } else {
                    s.type = id;
                }
                for (int i = 1; i < parts.length; i++) s.fields.add(parts[i]);
            }
            // Basic logging for debugging
            Log.d(TAG, "Parsed NMEA: " + s.toString());
            return s;
        } catch (Exception e) {
            Log.w(TAG, "Failed to parse NMEA: " + line, e);
            return null;
        }
    }

    public static Map<String, Object> quickSummary(Sentence s) {
        Map<String, Object> map = new HashMap<>();
        if (s == null) return map;
        map.put("talker", s.talker);
        map.put("type", s.type);
        map.put("fieldsCount", s.fields.size());
        return map;
    }

    public static Map<String, Double> extractLatLon(Sentence s) {
        Map<String, Double> map = new HashMap<>();
        if (s == null || s.type == null) return map;
        try {
            String t = s.type.toUpperCase();
            if (t.endsWith("GGA")) {
                // fields: [time, lat, NS, lon, EW, ...]
                if (s.fields.size() >= 5) {
                    String latStr = s.fields.get(1);
                    String ns = s.fields.get(2);
                    String lonStr = s.fields.get(3);
                    String ew = s.fields.get(4);
                    Double lat = nmeaToDecimal(latStr, ns);
                    Double lon = nmeaToDecimal(lonStr, ew);
                    if (lat != null && lon != null) {
                        map.put("lat", lat);
                        map.put("lon", lon);
                    }
                }
            } else if (t.endsWith("RMC")) {
                // fields: [time, status, lat, NS, lon, EW, ...]
                if (s.fields.size() >= 6) {
                    String latStr = s.fields.get(2);
                    String ns = s.fields.get(3);
                    String lonStr = s.fields.get(4);
                    String ew = s.fields.get(5);
                    Double lat = nmeaToDecimal(latStr, ns);
                    Double lon = nmeaToDecimal(lonStr, ew);
                    if (lat != null && lon != null) {
                        map.put("lat", lat);
                        map.put("lon", lon);
                    }
                }
            }
        } catch (Exception e) {
            Log.w(TAG, "extractLatLon error", e);
        }
        return map;
    }

    private static Double nmeaToDecimal(String val, String hemi) {
        if (val == null || val.isEmpty()) return null;
        try {
            double d = Double.parseDouble(val);
            int deg = (int) (d / 100);
            double min = d - deg * 100;
            double dec = deg + min / 60.0;
            if (hemi != null) {
                hemi = hemi.trim().toUpperCase();
                if (hemi.equals("S") || hemi.equals("W")) dec = -dec;
            }
            return dec;
        } catch (NumberFormatException e) {
            return null;
        }
    }
}

File: app/src/main/java/com/example/tugis3/gnss/nmea/NmeaParser.kt
package com.example.tugis3.gnss.nmea

import com.example.tugis3.gnss.model.FixType
import kotlin.math.abs

/**
 * Basit / hafif NMEA v2.x çözümleyici – yalnızca GGA ve RMC satırlarını işler.
 *
 * Üretim için: checksum doğrulama, daha fazla cümle (GSA / GSV), time / date birleşimi vb. eklenebilir.
 */
class NmeaParser {
    data class Parsed(
        val latDeg: Double? = null,
        val lonDeg: Double? = null,
        val heightEllipsoidal: Double? = null,
        val satellitesInUse: Int? = null,
        val satellitesVisible: Int? = null,
        val hdop: Double? = null,
        val vdop: Double? = null,
        val pdop: Double? = null,
        val fixType: FixType? = null,
        val raw: String
    )

    fun parse(line: String): Parsed? {
        if (!line.startsWith('$') || line.length < 10) return null
        // Checksum doğrulama (opsiyonel ama etkin)
        if (!validateChecksum(line)) return null
        val starIdx = line.indexOf('*')
        val body = if (starIdx > 0) line.substring(1, starIdx) else line.substring(1)
        val fields = body.split(',')
        if (fields.isEmpty()) return null
        return when (fields[0]) {
            "GPGGA", "GNGGA", "GAGGA" -> parseGGA(fields, line)
            "GPRMC", "GNRMC", "GARMC" -> parseRMC(fields, line)
            "GPGSA", "GNGSA", "GAGSA" -> parseGSA(fields, line)
            "GPGSV", "GNGSV", "GAGSV" -> parseGSV(fields, line)
            else -> null
        }
    }

    private fun parseGGA(f: List<String>, raw: String): Parsed? {
        // GGA formatı: $..GGA,utc,lat,NS,lon,EW,fix,sats,hdop,alt,M,geoid,M,...
        if (f.size < 10) return null
        val fixQuality = f.getOrNull(6)?.toIntOrNull() ?: 0
        val satCount = f.getOrNull(7)?.toIntOrNull()
        val hdop = f.getOrNull(8)?.toDoubleOrNull()
        val alt = f.getOrNull(9)?.toDoubleOrNull()
        val fix = mapFixQuality(fixQuality)
        val lat = dmToDeg(f.getOrNull(2), f.getOrNull(3))
        val lon = dmToDeg(f.getOrNull(4), f.getOrNull(5))
        return Parsed(
            latDeg = lat,
            lonDeg = lon,
            heightEllipsoidal = alt,
            satellitesInUse = satCount,
            hdop = hdop,
            fixType = fix,
            raw = raw
        )
    }

    private fun parseRMC(f: List<String>, raw: String): Parsed? {
        // RMC formatı: $..RMC,utc,status,lat,NS,lon,EW,sog,cog,date,...
        if (f.size < 10) return null
        val status = f.getOrNull(2)
        if (status != "A") { // A=Active, V=Void
            return Parsed(raw = raw, fixType = FixType.NO_FIX)
        }
        val lat = dmToDeg(f.getOrNull(3), f.getOrNull(4))
        val lon = dmToDeg(f.getOrNull(5), f.getOrNull(6))
        return Parsed(
            latDeg = lat,
            lonDeg = lon,
            raw = raw,
            // RMC kendi başına fix tipini ayırt etmez – SINGLE varsay.
            fixType = FixType.SINGLE
        )
    }

    private fun parseGSA(f: List<String>, raw: String): Parsed? {
        // $..GSA,Mode1,Mode2,Sat1,...,Sat12,PDOP,HDOP,VDOP*CS
        if (f.size < 17) return null
        val pdop = f.getOrNull(15)?.toDoubleOrNull()
        val hdop = f.getOrNull(16)?.toDoubleOrNull()
        val vdField = f.getOrNull(17)
        val vdop = vdField?.substringBefore('*')?.toDoubleOrNull()
        // Kullanılan uydu sayısı: sat alanlarından boş olmayanların sayısı
        val satInUse = f.subList(3, 15).count { it.isNotBlank() }
        // Mode2 -> 1: no fix, 2: 2D, 3: 3D (basit eşleme)
        val mode2 = f.getOrNull(2)
        val fixType = when (mode2) {
            "1" -> FixType.NO_FIX
            "2" -> FixType.SINGLE
            "3" -> FixType.SINGLE
            else -> null
        }
        return Parsed(
            satellitesInUse = if (satInUse > 0) satInUse else null,
            hdop = hdop,
            vdop = vdop,
            pdop = pdop,
            fixType = fixType,
            raw = raw
        )
    }

    private fun parseGSV(f: List<String>, raw: String): Parsed? {
        // $..GSV,totalMsgs,msgIdx,totalSats, ... (her cümlede 4 uyduya kadar)
        if (f.size < 4) return null
        val total = f.getOrNull(3)?.toIntOrNull()
        return Parsed(
            satellitesVisible = total,
            raw = raw
        )
    }

    private fun mapFixQuality(q: Int): FixType = when (q) {
        0 -> FixType.NO_FIX
        1 -> FixType.SINGLE
        2 -> FixType.DGPS
        3 -> FixType.PPP
        4 -> FixType.RTK_FIX
        5 -> FixType.RTK_FLOAT
        6 -> FixType.DGPS // Estimated (dead reckoning)
        7 -> FixType.MANUAL // Manual input mode
        8 -> FixType.DGPS // Simulator mode
        else -> FixType.SINGLE
    }

    private fun dmToDeg(dm: String?, hemi: String?): Double? {
        if (dm.isNullOrBlank() || hemi.isNullOrBlank()) return null
        val dot = dm.indexOf('.')
        if (dot < 0) return null
        // Latitude ddmm.mmmm, Longitude dddmm.mmmm
        val degLen = if (dot > 3) dot - 2 else dot - 2
        val degPart = dm.take(degLen).toIntOrNull() ?: return null
        val minPart = dm.drop(degLen).toDoubleOrNull() ?: return null
        var v = degPart + (minPart / 60.0)
        if (hemi.equals("S", true) || hemi.equals("W", true)) v = -abs(v)
        return v
    }

    private fun validateChecksum(line: String): Boolean {
        val star = line.indexOf('*')
        if (star < 0 || star + 3 > line.length) return false
        val provided = line.substring(star + 1).trim()
        val body = line.substring(1, star)
        var cs = 0
        for (c in body) cs = cs xor c.code
        val hex = cs.toString(16).uppercase().padStart(2, '0')
        return hex == provided.uppercase()
    }
}
